homeassistant/components/thethingsnetwork/entity.py

```diff
@@ -0,0 +1,203 @@
+
+from abc import ABC, abstractmethod
+import logging
+from typing import TYPE_CHECKING, Optional
+
+from ttn_client import TTNBaseValue, TTNSensorValue
+
+from homeassistant.core import callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+if TYPE_CHECKING:
+    from homeassistant.config_entries import ConfigEntry
+
+    from .coordinator import TTNCoordinator
+
+from .const import (
+    CONF_APP_ID,
+    OPTIONS_DEVICE_NAME,
+    OPTIONS_FIELD_ICON,
+    OPTIONS_FIELD_NAME,
+    OPTIONS_FIELD_PICTURE,
+    OPTIONS_FIELD_UNIT_MEASUREMENT,
+)
+from .entry_settings import TTN_EntrySettings
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TTN_Entity(CoordinatorEntity, Entity, ABC):
+    """Representation of a The Things Network Data Storage sensor."""
+
+    @staticmethod
+    def get_unique_id(device_id: str, field_id: str) -> str:
+        """Get unique_id which is derived from device_id and field_id."""
+        return f"{device_id}_{field_id}"
+
+    def __init__(
+        self,
+        entry: "ConfigEntry",
+        coordinator: "TTNCoordinator",
+        ttn_value: TTNSensorValue,
+    ) -> None:
+        """Initialize a The Things Network Data Storage sensor."""
+
+        self.__entry = entry
+        self._ttn_value = ttn_value
+
+        """Pass coordinator to CoordinatorEntity."""
+        super().__init__(coordinator, context=self.unique_id)
+
+        # Values from options
+        self._unit_of_measurement = None
+        self.__icon = None
+        self.__picture = None
+
+        self.__refresh_names()
+
+    # -----------------------------------------------------------#
+    # Methods to keep list of entities
+    #
+    # NOTE: the entity_registry helper cannot be used here as it
+    # returns instances created in the past, even if they are
+    # not longer available in TTN
+    # -----------------------------------------------------------#
+
+    async def async_added_to_hass(self) -> None:
+        """Remember added entity - see exits method below."""
+
+        await super().async_added_to_hass()
+        TTN_EntrySettings(self.__entry).get_entities()[self.unique_id] = self
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Remove entity from hass."""
+
+        await super().async_will_remove_from_hass()
+        TTN_EntrySettings(self.__entry).get_entities().pop(self.unique_id)
+
+    @staticmethod
+    def exits(entry: "ConfigEntry", device_id: str, field_id: str) -> bool:
+        """Check if an entry for this device/field already exists in HASS.
+
+        It is used to avoid creating duplicates while still allowing adding new devices/fields without restarting the adapter.
+        """
+        return (
+            TTN_Entity.get_unique_id(device_id, field_id)
+            in TTN_EntrySettings(entry).get_entities()
+        )
```

Comment:
Yea this is why I think moving the creation to the platforms would be beneficial as you can keep track of the created entities there (in the withings example, I keep track of all measurements I created sensors for, and if I get data for a point I haven't yet, I will create an entity)

Reply:
ok, I get the idea of keeping the list of created entities in the platform `async_setup_entry`. Cool, I did not like it keeping it under the entry data.